### ISSUE TITLE
Executable binaries in packages used for `build_requires` can fail to run on Linux

### ISSUE THREAD

**Issue by elizagamedev** [bincrafters/community#267]
### Description of Problem, Request, or Question
On Linux, certain prebuilt executable binaries can fail if the C++ runtime included on the system is too different from the version the prebuilt was built with. The reason for this is because `settings` does not include `"compiler"`. It might be worth considering adding a clause to include `"compiler"` to `settings` only if the package is targetting Linux, though it's possible macOS may be affected by a similar problem (unverified).

### Package Details (Include if Applicable)
* Package Name/Version: **cmake_installer**, **ninja_installer**, **nasm_installer**, **yasm_installer**, etc
* Operating System: **CentOS 6.9**
* Compiler+version: **gcc-4.4** and **gcc-6.2**

### Steps to reproduce (Include if Applicable)
Attempt to run the prebuilt cmake binary from cmake_installer on CentOS 6.9.

**Comment by SSE4** [comment]
yeah, that's long standing problem with conan, but actually it has no connection to the compiler version, it's actually glibc version incompatibility, so adding compiler to settings will not help at all.
conan currently doesn't model glibc version in settings and seems like there is no intention to do it.
the only solution I can imagine now is to build all build tools on some really really ancient distro to make sure it will work everywhere.
/cc @lasote @memsharded

**Comment by theirix** [comment]
We solved this problem for our internal infrastructure by building against an old version of centos. Unfortunately, old CentOS versions must have their own rpm repositories like http://vault.centos.org/6.6/os/x86_64 because official repository has only latest version (6.9). So it needs some efforts to build for an older version but it's up to choosing a right docker image. Conan recipe stays exactly same.

**Comment by SSE4** [comment]
alternate solution to look into https://github.com/wheybags/glibc_version_header

**Comment by SSE4** [comment]
CentOS 6.9 seems to be using glibc 2.12:
```
[root@3c4e2d14f992 /]# ldd --version
ldd (GNU libc) 2.12
Copyright (C) 2010 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```

**Comment by SSE4** [comment]
testing locally in **centos:6.9** container, without the fix:
```
[root@ca24cf60579b bin]# ./ninja --version
./ninja: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by ./ninja)
./ninja: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.18' not found (required by ./ninja)
./ninja: /usr/lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by ./ninja)
./ninja: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by ./ninja)
./ninja: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by ./ninja)
./ninja: /lib64/libc.so.6: version `GLIBC_2.15' not found (required by ./ninja)
```
```
[root@ca24cf60579b bin]# ./nasm  --version
./nasm: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by ./nasm)
```
```
[root@ca24cf60579b bin]# ./yasm --version
./yasm: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by ./yasm)
```
with glibc_version_header 2.12.2:
```
[root@ca24cf60579b bin]# ./nasm --version
NASM version 2.13.02 compiled on May 11 2018
```
```
[root@ca24cf60579b bin]# ./yasm --version
yasm 1.3.0
Compiled on May 11 2018.
Copyright (c) 2001-2014 Peter Johnson and other Yasm developers.
Run yasm --license for licensing overview and summary.
```
but ninja still needs libstdc++:
```
[root@ca24cf60579b bin]# ./ninja --version
./ninja: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by ./ninja)
./ninja: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.18' not found (required by ./ninja)
./ninja: /usr/lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by ./ninja)
./ninja: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by ./ninja)
```
so, something has to be done about libstdc++ version as well (and yes, **-static-libstdc++** doesn't help - it then needs memcpy from GLIBC 2.14)

**Comment by datalogics-kam** [comment]
Our project is needing boost package dependencies to be recompiled, and I solved the problem with `b2` not running somewhat by adding `--build boost_build` to the options on `conan install`.

Using CentOS 6.9.

**Comment by SSE4** [comment]
update, with glibc_version_header, x64 builds work fine, while x86 fails, investigating...

**Comment by SSE4** [comment]
now we finally have `glibc_version_header` working for both x86 and x64!
the following installers were updated:

- ninja_installer
- nasm_installer
- yasm_installer

other installers could be easily updated by adding the following lines:
```
    def build_requirements(self):
        if self.settings.os_build == 'Linux':
            self.build_requires('glibc_version_header/0.1@bincrafters/stable')
```
this will include proper glibc compatibility (`2.12.2` by default), and cause `libstdc++` and `libgcc` to be statically linked.
it makes possible to run installers on ancient binaries, such as CentOS.

**Comment by uilianries** [comment]
Awesome! Thanks @SSE4 !!
